### ISSUE TITLE
Word order in Arabic translations (Nammu 1.1.0)

### ISSUE THREAD

**Issue by EleanorRobson** [oracc/nammu#322]
Sorry it's taken me so long to get back to this. It's been a hectic week and I'm without a project administrator again :(

What do I do to switch script direction in the translation? It's not enough just to change to an Arabic font on my keyboard. Compare the screen shots attached: emacs-arabic.jpg shows what an Arabic translation ought to look like; nammu-arabic.jpg shows what I'm getting when I type the same text into Nammu.

![emacs-arabic](https://user-images.githubusercontent.com/4301009/42587115-adcbf7a2-8531-11e8-8e88-83d79d674946.jpg)
![nammu-arabic](https://user-images.githubusercontent.com/4301009/42587116-adf3d6fa-8531-11e8-87ee-68850cfade55.jpg)

Line 2 of the translation isn't bad, as at least the words are in the right order but in line 1, where I've followed the usual Arabic convention of typing ancient names in western script, the word order is completely confused. 

Am I missing a command or button, or does this need further work?

**Comment by sgrieve** [comment]
Hi @EleanorRobson Thanks for the feedback. In theory the Arabic text should be automatically detected, so clearly something is going wrong somewhere. We will investigate and get a new test version to you soon.

Thanks!

**Comment by sgrieve** [comment]
Doing some preliminary testing and it seems that the automatic detection of Arabic translations fails when using the split screen view. I will investigate why this is and attempt to find a fix.

**Comment by EleanorRobson** [comment]
Great, thank you! Split screen is the default translation mode, at least for me, as I need to see what I'm translating :)

**Comment by raquel-ucl** [comment]
Hi @EleanorRobson. @sgrieve has made a new version of the code with a fix for your problem. He will be working on refactoring the fix next week to make it more performant since it became a little bit slow when using the mac trackpad. In any case you can test the functionality for the split view to edit Arabic translations already. Please, let us know if this works for you now or you have any other comments.

You can find the new version of the jar is here: https://github.com/oracc/nammu/releases/tag/untagged-8e81d591d83835d819fb

**Comment by EleanorRobson** [comment]
It's *incredibly * slow on my MacBook Air. It does work, just about, but nowhere near fast enough for a live demo yet.

And may I ask where we're at with the top-level website search? Will there be a demo-able version ready for Wednesday afternoon?

Most gratefully,


e

**Comment by sgrieve** [comment]
H @EleanorRobson Just to keep you up to date, we are working on this issue now and will give you an update this afternoon on our progress.

**Comment by raquel-ucl** [comment]
We have just made a quick fix to isolate the problem, although we had to disable syntax highlighting for now. @EleanorRobson can you get this JAR, and let us know if the speed and functionality are alright? https://github.com/oracc/nammu/releases/tag/untagged-c7d4a8995b4633da714b

If it is, we'll need to spend some more time fixing the syntax highlighter. For your demo, we could do a quicker fix to use a similar colouring pattern than the Emacs plugin was using, without using pyoracc.

**Comment by EleanorRobson** [comment]
Yes, that's faster now but there's still the same word order problem when I switch fonts. If you compare this screen shot with the Emacs screen shot in my initial report, the words على أمر should be at the start (right) of the line. Nammu seems to be treating each sequence of words in Arabic (where the letters are correctly running R-L) as a single word in a L-R sentence. 
![nammu-1-1-2](https://user-images.githubusercontent.com/4301009/42758148-c0835b22-8902-11e8-96c6-1ef79833d8c7.jpg)

**Comment by sgrieve** [comment]
Hi @EleanorRobson,

We've been working on this all day and have found that it is not possible to correctly set the text direction in Java in lines where there is a mix of left to right and right to left scripts.

The solution we are proposing is to redesign the split window feature, so that the bottom pane contains the translation and the upper pane contains the rest of the atf. This will allow us to render the mixed Arabic and English translation lines correctly.

We will work all day tomorrow on this and keep you informed of our progress.

**Comment by EleanorRobson** [comment]
OK, thanks for letting me know

**Comment by sgrieve** [comment]
Hi @EleanorRobson,

We have just created a new release with the above solution. Not everything is implemented, and the underlying code is a bit messy, but hopefully it will get the job done for your demo tomorrow. Please let us know if it is all correct.

We have added another menu option in `window` called `Toggle Arabic translation window`. This creates the split pane view, but the lower pane is configured to display Arabic text (with in line English words) correctly.

If you load a file containing an existing Arabic translation, detected from the `@translation` line, the translation will open in the lower pane.

If you open a file without an Arabic translation, it will open by default in a single pane, and translation mode can be enabled using the option in the `window` menu. This will open the bottom translation pane, ready for you to begin translating. Once you have finished translating, pressing save will join the new translation with the rest of the atf file.

The release can be found here: https://github.com/oracc/nammu/releases/tag/untagged-b075d3101499a8e47f66

**Comment by EleanorRobson** [comment]
Thank you both! That's more than good enough for tomorrow. 

You can see my tiny test-corpus here: [http://build-oracc.museum.upenn.edu/cams/arabic/corpus](http://build-oracc.museum.upenn.edu/cams/arabic/corpus) — and it turns out that the Oracc pager has the same word order issues :)

I'm REALLY grateful for your speedy responsiveness!

**Comment by raquel-ucl** [comment]
I'm very happy to hear it's working for the demo today :) We are trying to make a new version re-enabling syntax highlight. It if works easily we'll upload a new jar before lunch time. Otherwise we'll just work on it for the next release once we sort out how is best to deal with RTL translations.

**Comment by sgrieve** [comment]
Hi @EleanorRobson 

We have reenabled syntax highlighting for the main editor pane. So when working in the new Arabic translation mode the upper pane is syntax highlighted as before, and the Arabic translation has the same font, but no highlighting.

When we test it, everything works nice and fast, so hopefully it will be ok for you as well.

Heres the link: https://github.com/oracc/nammu/releases/tag/untagged-f74e2a69f34d2c0ad9ea

**Comment by EleanorRobson** [comment]
Thanks so much, both of you! I downloaded this version 10 minutes before the workshop, and successfully live-edited and uploaded a text with Arabic translation. Even my Iraqi colleagues approve :)

**Comment by sgrieve** [comment]
Thats fantastic news!
